### PR TITLE
Added support of native menus in desktop mode #5503

### PR DESCRIPTION
**Pull request by nikhil-mohite**

### CODE REVIEW

File: web/pgadmin/browser/__init__.py [hunk 1]
```diff
@@ -683,7 +684,8 @@ def utils():
             qt_default_placeholder=QT_DEFAULT_PLACEHOLDER,
             vw_edt_default_placeholder=VW_EDT_DEFAULT_PLACEHOLDER,
             enable_psql=config.ENABLE_PSQL,
-            pgadmin_server_locale=default_locale
+            pgadmin_server_locale=default_locale,
+            is_nwjs='true' if current_app.PGADMIN_RUNTIME else 'false',
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/__init__.py:688]: We don't need this. We already have a JS function `getBrowser` for this. You can:
```
    let {name: browser} = getBrowser();
    if(browser == 'Nwjs') {
```

File: web/pgadmin/browser/static/js/browser.js [hunk 2]
```diff
@@ -366,6 +372,74 @@ define('pgadmin.browser', [
       scripts[n].push({'name': m, 'path': p, loaded: false});
     },
     masterpass_callback_queue: [],
+    getMenuList: function(name, item, d, skipDisabled=false) {
+      let obj = this;
+      let category = {
+        'common': []
+      };
+      for(let _key of Object.keys(obj.menus[name][d._type])){
+        if(obj.menus[name][d._type][_key].category in this.menu_categories) {
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/static/js/browser.js:381]: Create a var for `obj.menus[name][d._type][_key].category` and use it to simplify code.

File: web/pgadmin/browser/static/js/new_menus.js [hunk 3]
```diff
@@ -0,0 +1,236 @@
+import _ from 'lodash';
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/static/js/new_menus.js:1]: `new_menu.js`

File: web/pgadmin/browser/static/js/new_menus.js [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+      name: label,
+      /* icon: this.icon || this.module && (this.module.type), */
+      disabled: is_disabled,
+      callback: () => { this.contextMenuCallback(self, item); },
+      ...(sub_ctx_item && Object.keys(sub_ctx_item).length > 0) && { items: sub_ctx_item }
+    };
+  }
+
+
+  change_checked() {
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/static/js/new_menus.js:163]: Cleanup.

File: web/pgadmin/browser/static/js/new_menus.js [hunk 5]
```diff
@@ -0,0 +1,236 @@
+import _ from 'lodash';
+import * as pgadminUtils from 'sources/utils';
+import gettext from 'sources/gettext';
+import pgAdmin from 'sources/pgadmin';
+
+class Menu {
+  constructor() {
+    this.primaryMenus = [
+      { label: gettext('File'), menu: 'file', id: '#mnu_file', index: 0, isMenu: true, allowSubmenus: true, addSepratior: true },
+      { lebel: gettext('Management'), menu: 'management', id: '#mnu_management', isMenu: false, allowSubmenus: false, addSepratior: true },
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/static/js/new_menus.js:10]: Remove management. It is no where used.

File: web/pgadmin/browser/static/js/new_menus.js [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    if (this.module && _.isFunction(this.module[this.enable])) {
+      return !(this.module[this.enable]).apply(this.module, [node, item, this.data]);
+    }
+
+    return false;
+  }
+}
+
+
+function createMenu() {
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/static/js/new_menus.js:212]: Move to this Menu class with name `static function create`.
And return the menuObj as return value.

Caller can call as `const menuObj = Menu.create()`

File: web/pgadmin/browser/static/js/new_menus.js [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+function getContextMenu(menu, item, node) {
+  let menuObj = new Menu();
+  return menuObj.getContextMenus(menu, item, node);
+
+}
+
+module.exports = {
+  newMenu: Menu,
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/static/js/new_menus.js:231]: Use ES6.

File: web/pgadmin/browser/static/js/new_menus.js [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+function createMenu() {
+  let menuObj = new Menu();
+  menuObj.primaryMenus.forEach((_menu) => {
+    menuObj.generate(_menu);
+  });
+}
+
+function createMenuItem(menuItems) {
+  let menuObj = new Menu();
+  menuObj.generateObjectMenu(menuItems);
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/static/js/new_menus.js:221]: Why create object menu here?

File: web/pgadmin/browser/static/js/node.js [hunk 9]
```diff
@@ -112,6 +112,7 @@ define('pgadmin.browser.node', [
         priority: 2,
         label: gettext('Refresh...'),
         icon: 'fa fa-sync-alt',
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/static/js/node.js:114]: may be its time we clean up `icon` key. It is not used, and will not be used.

File: web/pgadmin/browser/templates/browser/index.html [hunk 10]
```diff
@@ -98,6 +98,7 @@
     <div class="row"><div class="col-12 pg-sp-text">{{ _('Loading {0} v{1}...').format(config.APP_NAME, config.APP_VERSION) }}</div></div>
   </div>
 </div>
+{% if is_nwjs == 'false' %}
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/templates/browser/index.html:101]: You can directly check env PGADMIN_RUNTIME

File: web/pgadmin/browser/templates/browser/js/utils.js [hunk 11]
```diff
@@ -104,7 +104,9 @@ define('pgadmin.browser.utils',
         {% for key in ('File', 'Edit', 'Object' 'Tools', 'Management', 'Help') %}
         obj.add_menus({{ MENU_ITEMS(key, current_app.menu_items['%s_items' % key.lower()])}});
         {% endfor %}
-        obj.create_menus();
+        if('{{is_nwjs}}' != 'true') {
```
Comment (adityatoshniwal) [inline web/pgadmin/browser/templates/browser/js/utils.js:107]: use getBrowser

File: web/pgadmin/static/bundle/app.js [hunk 12]
```diff
@@ -39,5 +41,13 @@ define('app', [
   initializeModules(pgAdmin.Tools);
 
   // create menus after all modules are initialized.
-  pgAdmin.Browser.create_menus();
+  let {name: browser} = getBrowser();
+  if(browser == 'Nwjs') {
```
Comment (adityatoshniwal) [inline web/pgadmin/static/bundle/app.js:45]: This should be handled inside `create_menus` function. No change needed here.

**Review by adityatoshniwal — CHANGES_REQUESTED** [review]
Hi @nikhil-mohite Did a code walk through and reporting some suggestions here. Please check.

File: web/pgadmin/browser/templates/browser/js/utils.js (hunk 11 above)
Comment (nikhil-mohite) [reply web/pgadmin/browser/templates/browser/js/utils.js:107]: updated with `current_app.PGADMIN_RUNTIME ` check

File: runtime/src/js/pgadmin.js [hunk 13]
```diff
@@ -352,6 +368,236 @@ splashWindow.on('loaded', function() {
...
+  }
+
+}
+
+function addMacMenus(pgBrowser, menuGroup, menuItems) {
+  if (menuGroup.menu == 'file' && menuGroup.isMenu && platform() === 'darwin') {
+    var rootMenu = nativeMenu.items[0].submenu;
+    let indx = 0;
+    for (let item_key of Object.keys(menuItems)) {
+      menuItems[item_key].forEach((menuItem) => {
```
[the remark this answers is not in the document]
Comment (nikhil-mohite) [reply runtime/src/js/pgadmin.js:458]: created a common function to get a sub-menu list, but for macOS, we use `rootMenu.insert` to add the menu at a specific location so created a separate function for macOS.

File: web/pgadmin/browser/static/js/new_menus.js (hunk 6 above)
Comment (nikhil-mohite) [reply web/pgadmin/browser/static/js/new_menus.js:212]: Done

**Pull request closed without merging**